src/daft-catalog/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+use std::{collections::HashMap, default, sync::Arc};
+
+use daft_plan::LogicalPlanBuilder;
+use data_catalog::DataCatalog;
+
+mod data_catalog;
+mod data_catalog_table;
+mod errors;
+
+#[cfg(feature = "python")]
+pub mod python;
+
+use errors::Error;
+use once_cell::sync::Lazy;
+
+/// Global singleton DaftMetaCatalog, lazily initialed from the environment
+pub(crate) static GLOBAL_DAFT_META_CATALOG: Lazy<DaftMetaCatalog> =
+    Lazy::new(DaftMetaCatalog::new_from_env);
+
+/// The [`DaftMetaCatalog`] is a catalog of [`DataCatalog`] implementations
+///
+/// Users of Daft can register various [`DataCatalog`] with Daft, enabling
+/// discovery of tables across various [`DataCatalog`] implementations.
+pub struct DaftMetaCatalog {
+    default_data_catalog: Option<Arc<dyn DataCatalog>>,
+    named_data_catalogs: HashMap<String, Arc<dyn DataCatalog>>,
+    view_catalog: HashMap<String, LogicalPlanBuilder>,
```

Comment:
Consider using interior mutability (e.g., Mutex) for `named_data_catalogs` and `view_catalog`. This would allow modifications through shared references, enabling use of `Arc<DaftMetaCatalog>` without &mut self in register methods.

Reply:
Great points.... I was trying to find a way around this also specifically because I wanted to enable lazy materialization of PyIceberg objects inside of the GlueCatalog (if we receive information from Glue that a given table is an `"iceberg"` table, then we can lazily initialize the PyIceberg Glue client to help us interface with it)

I'm not actively working on this PR yet until perhaps next week, but will work in interior mutability here which is needed to make these happen.